[REF/#386] Custom Calendar 구현

angryPodo opened the pull request:
## Related issue 🛠
- closed #386 

## Work Description ✏️
- `kizitonwose/Calendar` 의존성 제거
- `Calendar` 컴포넌트 구현

## Screenshot 📸

https://github.com/user-attachments/assets/73af0d74-5a84-4543-93a3-ff7d1e688940


## To Reviewers 📢
여태까지 Figma와 맞지 않는 부분을 보면서 분했었는데, 날잡고 고치자 마음먹어서 바로 진행했습니다ㅎㅎ
라이브러리를 많이 참고했고, 경량화에 집중했다고 봐주시면 됩니다ㅎㅎ

angryPodo committed:
[REF/#386] Calendar 리펙토링

Hyobeen-Park commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/DaysOfWeekTitle.kt:
```diff
@@ -48,3 +56,15 @@ internal fun DaysOfWeekTitle(
+
+@Preview
+@Composable
+internal fun DaysOfWeekTitlePreview() {
```
internal은 의도인가요?

angryPodo replied:
엥 아뇨 잘못작성했네요😅

Hyobeen-Park commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/util/MonthData.kt:
```diff
@@ -0,0 +1,51 @@
+            month -> DayPosition.MonthDate
+            previousMonth -> DayPosition.InDate
+            nextMonth -> DayPosition.OutDate
+            else -> throw Exception()
```
에러 던진거 처리해주는 코드가 있나요???

angryPodo replied:
아뇨 의도적으로 앱 크래시 내는게 맞습니다. 왜냐면 지금 이 로직이 실패했다는건 "절대 실행될 수 없는 부분"이라서요. 네트워크 오류처럼 예상된 예외가 아니라 논리결함이라서 어설프게 앱을 동작시키기 보다 빠르게 중단시키는 겁니다. 그래야 나중에 크래시 리포트 툴에서 어떤 조건에서 발생했는지 추적이 가능하니까요. 또한 그냥 에러를 감싸서 크래시를 내지않고 실행시키면 데이터가 꼬이거나 예측 불가한 문제가 연쇄될 수 있으니ㅎㅎ
예로들면 `pickMultipleVisulMedia`에서 최소 선택 조건 수 > 1 을 만족하지 않으면 앱을 크래시 내는것과 같습니다.

angryPodo replied:
대신 에러 로깅을 더 추가하겠습니다 :)

Hyobeen-Park reviewed (approved):
아니아니 세상에 이걸 스프린트 중에 해냈다구요?!???? 역시 짱이야

nhyeonii commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/HorizontalCalendar.kt:
```diff
@@ -0,0 +1,112 @@
+    lazyListState = lazyListState,
+    snapPosition = SnapPosition.Start
+) {
+    override fun calculateApproachOffset(velocity: Float, decayOffset: Float): Float = 0f
```
항상 0f를 반환하는건 애니메이션 없이 바로 스냅 될 수 있도록 하기 위해서 일까요 ?!

angryPodo replied:
아니요 애니메이션은 관계없습니다. 기본 `flingBehavior`은 스크롤이 제약없이 됩니다. 하나의 페이지씩 넘어가지 않아서 커스텀을 만든겁니다

nhyeonii reviewed (approved):
우와 커스텀 캘린더 ... !!! 짱 신기하네요.. 진짜 대단하다이 ..!!! 고생하셨습니다 !!!

nahy-512 commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/model/DateExtensions.kt:
p2
Extenstions이 model 패키지에 있는 이유가 있을까요?
저희 core에서는 폴더를 따로 분리했던 거 같은데, 맥락 상으로는 util에 있는 편이 조금 더 잘 어울리지 않을까 하는 생각도 들어서요

angryPodo replied:
음,,,그렇긴하네요. 요건 패키징 나누겠습니다

nahy-512 commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/model/DayPosition.kt:
```diff
@@ -0,0 +1,7 @@
+package com.hilingual.presentation.home.component.calendar.model
+
+internal enum class DayPosition {
+    InDate,
+    MonthDate,
+    OutDate
```
p3
InDate는 이전 달의 날짜, OutDate는 다음 달의 날짜로 파악했는데, 두개를 구분하신 이유가 있으실까요?
이번 달에 속하지 않은 날짜 처리 방침은 똑같은 걸로 알고 있어서, 추후 확장성을 위해 구분하신 건지 궁금합니다!

angryPodo replied:
장기적으로 분리된 상태를 유지하는게 더 유리하다고 생각합니다. 지난달 날짜는 회색으로, 다음 달 날짜는 아예 선택불가능하게 라는 변경이 생기면 상태를 또 나눠야겠죠. 그떄가서 분리하는것보다 지금 구분하는게 더 좋다고 생각합니다. 이정도의 우선적인 분리는 큰 오버 엔지니어링이 아니라고 생각합니다

nahy-512 commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/util/MonthData.kt:
p3
저는 data class라면 보통 model에 넣는 클래스다..! 라는 생각이 있었는데요.
이런 식으로 data class 안에 변수 선언 + 로직적인 코드를 넣게 된다면 util로 볼 수도 있겠군요

angryPodo replied:
오호라...보통 model에 넣는 클래스는 맞지만 만약 그런 이유로만 사용하셨다면 `data class`에 대해서 더 이해해보면 좋겠습니다.
왜 model에서 `data class`를 사용할까요? 뭐가 편리해서? 뭐가 좋길래? 뭐를 제공해주길래 사용하지? 라는 물음을 던져보시고 알아보십쇼ㅎㅎ

nahy-512 commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/state/CalendarState.kt:
```diff
@@ -0,0 +1,56 @@
+
+    val monthIndicesCount: Int = getMonthIndicesCount(startMonth, CALENDAR_END_MONTH)
+
+    suspend fun scrollToMonth(month: YearMonth) {
+        listState.scrollToItem(getMonthIndex(startMonth, month))
+    }
+
+    suspend fun animateScrollToMonth(month: YearMonth) {
+        listState.animateScrollToItem(getMonthIndex(startMonth, month))
+    }
```
p3
오호.. 피커로 달 이동할 때랑 스와이프로 이동할 때랑 애니메이션을 구분하셨군요

angryPodo replied:
음 그렇다기는 보단..애니메이션을 쓰는 스크롤과 아닌 스크롤을 구분했습니다. 만약에 사용자 층에서 버벅거림이 많이 발생한다면 상위 메서드로 전환해야겠죠?

nahy-512 commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/util/MonthData.kt:
```diff
@@ -0,0 +1,51 @@
+    val firstDay = month.atStartOfMonth()
+    val inDays = firstDayOfWeek.daysUntil(firstDay.dayOfWeek)
+    val outDays = (inDays + month.lengthOfMonth()).let { inAndMonthDays ->
+        if (inAndMonthDays % 7 != 0) 7 - (inAndMonthDays % 7) else 0
```
p2
7이 일주일을 나타내는 값으로 파악했어요. 어떻게 보면 매직넘버일 것도 같은데, 너무 하드코딩이지는 않을까 하는 걱정이 드네요.
민재님은 이런 경우에도 상수로 분리하지 않고 숫자로 바로 넣어주는 방식을 선호하시나요??

angryPodo replied:
뭔가 알고리즘처럼 사용되는 부분에서 7을 그냥 쓴다면 컨텍스트 파악이 힘들겠지만, 지금은 "날짜"라는 상황에서 쓰이는 숫자 7이니 바로 파악이 가능하다고 생각했습니다. 상수로 추출한다고 하면 `DAYS_IN_WEEK`같이 추출할텐데 `if (inAndMonthDays % DAYS_IN_WEEK != 0)` 이게 더 파악하기 힘들다고 생각해요

nahy-512 reviewed (approved):
캘린더 의존성 다 제거된 거 보고 놀라서 들어왔습니다ㅎ.ㅎ
캘린더 초기 구현 시에 좋은 라이브러리 찾아서 수월하게 개발하셨다고 말씀하셨던 게 기억에 남는데, 그럼에도 커스텀 불가능한 영역에 대해 계속 신경쓰고 계셨군요.
기존 라이브러리 코드를 이런 식으로 앱 내에 맞게 커스텀할 수도 있네요. 내부 코드는 제가 더 공부해야 할 거 같지만,, 덕분에 많이 배워갑니다!

angryPodo committed:
[MOD/#386] Refactor calendar util and model files

- Move `DateExtensions.kt` from model to util package.
- Remove unused imports in `MonthData.kt`.
- Replace `Exception()` with `error()` for unexpected month date in `MonthData.kt`.
- Change internal preview function to private in `DaysOfWeekTitle.kt`.

Daljyeong commented on presentation/home/src/main/java/com/hilingual/presentation/home/component/calendar/state/CalendarState.kt:
```diff
@@ -0,0 +1,56 @@
+import java.time.DayOfWeek
+import java.time.YearMonth
+
+private val CALENDAR_START_MONTH = YearMonth.of(2025, 1)
+private val CALENDAR_END_MONTH = YearMonth.of(2100, 12)
```
p3
혹시 이 부분은 기명에는 없고 구두로 확정된 부분이었을까요??

angryPodo replied:
아뇨 기명에 있어요 코멘트로 논의 오고갔어요

Daljyeong reviewed (approved):
넘나리 대단..,,.. 덕분에 커스텀 캘린더 코드 확인하구 가요..!!
수고하셨습니당 🚀🚀
